Limit mempool size

drgora opened the pull request:
This PR adds a new CLI parameter, `-maxmempool`, that allows setting an upper limit on the size in bytes of the content of the mempool.
In order to respect such limit, transactions or certificates can be rejected/evicted according to context specific rules.

drgora committed:
Add mempool_size_limit Python test

This also includes a snapshot of a given chain, used to avoid the time consuming creation of certificate proofs on each
run.

drgora committed:
Implement mempool size limitation

drgora committed:
Make mempool a std::unique_ptr, instead of a value.

This allows avoiding static init, potentially losing user defined parameters parsed during Zend startup.

ptagl commented on src/gtest/test_sidechain_to_mempool.cpp:
```diff
@@ -241,58 +242,58 @@ TEST_F(SidechainsInMempoolTestSuite, hasSidechainCreationTxTest) {
 }
 
 TEST_F(SidechainsInMempoolTestSuite, ScAndFwdsAndBtrInMempool_ScNonRecursiveRemoval) {
-    // Associated scenario: Sidechain creation and some fwds and btr are in mempool.
```
I guess this is one too many replacements here! 😄

drgora replied:
Oh, right, a minor flaw in the otherwise brilliant `replace_all` idea :)

ptagl commented on src/gtest/test_sidechain_to_mempool.cpp:
```diff
@@ -387,80 +388,80 @@ TEST_F(SidechainsInMempoolTestSuite, ScAndFwdsAndBtrInMempool_ScRecursiveRemoval
-    // Associated scenario: upon block disconnections fwds and btr have entered into mempool.
-    // While unmounting block containing scCreation, scCreation cannot make to mempool. fwds and btr must me purged
+    // Associated scenario: upon block disconnections fwds and btr have entered into mempool->
+    // While unmounting block containing scCreation, scCreation cannot make to mempool-> fwds and btr must me purged
```
Also here we should probably keep "."!

drgora replied:
Done

ptagl commented on src/gtest/test_sidechain_to_mempool.cpp:
```diff
@@ -387,80 +388,80 @@ TEST_F(SidechainsInMempoolTestSuite, ScAndFwdsAndBtrInMempool_ScRecursiveRemoval
 
 TEST_F(SidechainsInMempoolTestSuite, ScAndFwdsAndBtrInMempool_FwdRecursiveRemoval) {
-    // Associated scenario: upon block disconnections a fwd cannot make to mempool.
+    // Associated scenario: upon block disconnections a fwd cannot make to mempool->
```
```suggestion
    // Associated scenario: upon block disconnections a fwd cannot make to mempool.
```

drgora replied:
Done

ptagl commented on src/gtest/test_sidechain_to_mempool.cpp:
```diff
@@ -387,80 +388,80 @@ TEST_F(SidechainsInMempoolTestSuite, ScAndFwdsAndBtrInMempool_ScRecursiveRemoval
 
 TEST_F(SidechainsInMempoolTestSuite, ScAndFwdsAndBtrInMempool_BtrRecursiveRemoval) {
-    // Associated scenario: upon block disconnections a btr cannot make to mempool.
+    // Associated scenario: upon block disconnections a btr cannot make to mempool->
```
```suggestion
    // Associated scenario: upon block disconnections a btr cannot make to mempool.
```

drgora replied:
Done

ptagl commented on src/gtest/test_sidechain_to_mempool.cpp:
```diff
@@ -831,12 +832,12 @@ TEST_F(SidechainsInMempoolTestSuite, UnconfirmedCsw_LargerThanSidechainBalanceAr
+    EXPECT_TRUE(mempool->addUnchecked(cswTx.GetHash(), cswEntry));
 
-    // Sidechain State is Ceased and there is no Sidechain balance conflicts in the mempool. No removed Txs and Certs expected.
+    // Sidechain State is Ceased and there is no Sidechain balance conflicts in the mempool-> No removed Txs and Certs expected.
```
```suggestion
    // Sidechain State is Ceased and there is no Sidechain balance conflicts in the mempool. No removed Txs and Certs expected.
```

drgora replied:
Done

ptagl commented on src/gtest/test_sidechain_to_mempool.cpp:
```diff
@@ -881,14 +882,14 @@ TEST_F(SidechainsInMempoolTestSuite, UnconfirmedCswForAliveSidechainsAreRemovedF
-    // Sidechain State is Ceased and there is no Sidechain balance conflicts in the mempool. No removed Txs and Certs expected.
-    mempool.removeStaleTransactions(&sidechainsView, removedTxs, removedCerts);
-    mempool.removeOutOfScBalanceCsw(&sidechainsView, removedTxs, removedCerts);
+    // Sidechain State is Ceased and there is no Sidechain balance conflicts in the mempool-> No removed Txs and Certs expected.
```
```suggestion
    // Sidechain State is Ceased and there is no Sidechain balance conflicts in the mempool. No removed Txs and Certs expected.
```

drgora replied:
Done

ptagl commented on src/init.cpp:
```diff
@@ -1966,7 +1974,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 
     // Start the thread that notifies listeners of transactions that have been
-    // recently added to the mempool.
+    // recently added to the mempool->
```
```suggestion
    // recently added to the mempool.
```

drgora replied:
Done

ptagl commented on src/rpc/blockchain.cpp:
```diff
@@ -2675,15 +2675,15 @@ UniValue getcertmaturityinfo(const UniValue& params, bool fHelp)
 
 /**
- * @brief Removes any transaction from the mempool.
+ * @brief Removes any transaction from the mempool->
```
```suggestion
 * @brief Removes any transaction from the mempool.
```

drgora replied:
Done

ptagl commented on src/rpc/blockchain.cpp:
```diff
@@ -2675,15 +2675,15 @@ UniValue getcertmaturityinfo(const UniValue& params, bool fHelp)
         throw runtime_error(
             "clearmempool\n"
-            "\nRemoves any transaction and certificate from the mempool. Wallets are NOT synchronized.\n"
+            "\nRemoves any transaction and certificate from the mempool-> Wallets are NOT synchronized.\n"
```
```suggestion
            "\nRemoves any transaction and certificate from the mempool. Wallets are NOT synchronized.\n"
```

drgora replied:
Done

ptagl commented on src/main.cpp:
```diff
@@ -1622,7 +1625,7 @@ MempoolReturnValue AcceptTxToMemoryPool(CTxMemPool& pool, CValidationState &stat
 
             // We pass pcoinsTip to IsScTxApplicableToState because we want to validate the fees against the last certificate
-            // in the blockchain, and not against certificates in the mempool.
+            // in the blockchain, and not against certificates in the mempool->
```
```suggestion
            // in the blockchain, and not against certificates in the mempool.
```

drgora replied:
Done

ptagl commented on src/main.cpp:
```diff
@@ -1795,24 +1798,26 @@ MempoolReturnValue AcceptTxToMemoryPool(CTxMemPool& pool, CValidationState &stat
-        pool.addUnchecked(hash, entry, !IsInitialBlockDownload());
+        if (!pool->addUnchecked(hash, entry, !IsInitialBlockDownload())) {
+            return MempoolReturnValue::MEMPOOL_FULL;
+        }
```
I've noted that in case of failures of the `addUnchecked()` function inside `AcceptCertificateToMemoryPool()` we print a debug line, would it be useful also here for standard transactions?

drgora replied:
Makes sense, done

ptagl commented on src/main.cpp:
```diff
@@ -4343,16 +4348,16 @@ bool static ConnectTip(CValidationState &state, CBlockIndex *pindexNew, CBlock *
     LogPrint("bench", "  - Writing chainstate: %.2fms [%.2fs]\n", (nTime5 - nTime4) * 0.001, nTimeChainState * 0.000001);
 
-    // Remove conflicting transactions from the mempool.
+    // Remove conflicting transactions from the mempool->
```
```suggestion
    // Remove conflicting transactions from the mempool.
```

drgora replied:
Done

ptagl commented on src/main.cpp:
```diff
@@ -6694,33 +6699,33 @@ void static ProcessGetData(CNode* pfrom)
         // do that because they want to know about (and store and rebroadcast and
         // risk analyze) the dependencies of transactions relevant to them, without
-        // having to download the entire memory pool.
+        // having to download the entire memory pool->
```
```suggestion
        // having to download the entire memory pool.
```

drgora replied:
Done

a-petrini commented on qa/rpc-tests/mempool_size_limit.py:
```diff
@@ -0,0 +1,593 @@
+from test_framework.util import assert_equal, initialize_chain_clean, \
+    start_nodes, sync_blocks, sync_mempools, disconnect_nodes, connect_nodes_bi, mark_logs,\
+    get_epoch_data, wait_bitcoinds, stop_nodes, \
+    swap_bytes, wait_and_assert_operationid_status
```
There are a few unused imports: `disconnect_nodes`, `wait_bitcoinds`, `stop_nodes`, `wait_and_assert_operationid_status`

drgora replied:
Good point, done

a-petrini replied:
Nice, thanks!

ptagl commented on src/main.cpp:
```diff
@@ -1469,23 +1469,26 @@ MempoolReturnValue AcceptCertificateToMemoryPool(CTxMemPool& pool, CValidationSt
 
         // Store transaction in memory
-        pool.addUnchecked(certHash, entry, !IsInitialBlockDownload());
+        if (!pool->addUnchecked(certHash, entry, !IsInitialBlockDownload())) {
```
I wonder if it's possible to make this check (also) earlier in this function, just to avoid making expensive checks when the mempool is full and unable to include it.

drgora replied:
I believe this might be an interesting follow up improvement, but requires some design plan and modifications to the current code (for example to have a dry-run `trimToSize` which does not actually remove things from mempool.

a-petrini commented on src/main.cpp:
```diff
@@ -1795,7 +1798,9 @@ MempoolReturnValue AcceptTxToMemoryPool(CTxMemPool& pool, CValidationState &stat
 
-        pool.addUnchecked(hash, entry, !IsInitialBlockDownload());
+        if (!pool.addUnchecked(hash, entry, !IsInitialBlockDownload())) {
+            return MempoolReturnValue::MEMPOOL_FULL;
```
Just for symmetry with the `addUnchecked()` for certs, it would be nice to add an informative `LogPrint()`

a-petrini replied:
Thanks!

drgora commented on src/txmempool.cpp:
```diff
@@ -2009,6 +2063,80 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
+    for (const CMemPoolEntry* r: to_be_removed) {
+        std::list<CTransaction> removed_txs;
+        std::list<CScCertificate> removed_certs;
+        if (certificate) {
```
This condition is wrong, this check should be on `r`. Will fix this with the next push, sorry about that.

a-petrini commented on src/wallet/wallet.h:
```diff
@@ -1117,7 +1117,7 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
 
     static CFeeRate minTxFee;
-    static CAmount GetMinimumFee(unsigned int nTxBytes, unsigned int nConfirmTarget, const CTxMemPool& pool);
+    static CAmount GetMinimumFee(unsigned int nTxBytes, unsigned int nConfirmTarget, const std::shared_ptr<CTxMemPool> pool);
```
To preserve the `const`-ness of `pool` variable which was present in the original function signature, I suggest 
`std::shared_ptr<const CTxMemPool> pool`

titusen replied:
Unnecessary copy constructor call. Please, pass by reference.

a-petrini replied:
Switching to `std::unique_ptr` and reverting to pass a const reference made my first comment obsolete

a-petrini commented on src/main.cpp:
```diff
@@ -6699,33 +6699,33 @@ void static ProcessGetData(CNode* pfrom)
 }
 
-void ProcessMempoolMsg(const CTxMemPool& pool, CNode* pfrom)
+void ProcessMempoolMsg(const std::shared_ptr<CTxMemPool> pool, CNode* pfrom)
```
Similarly to a following comment, to preserve the `const`-ness, I would suggest `std::shared_ptr<const CTxMemPool> pool`

a-petrini replied:
As before, latest changes made this comment obsolete

ptagl commented on src/txmempool.cpp:
```diff
@@ -2009,6 +2063,80 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
+    const CCertificateMemPoolEntry* certificate = dynamic_cast<const CCertificateMemPoolEntry*>(entry);
+    if (certificate && totalTxSize > m_max_size / 2) {
+        LogPrint("mempool", "%s():%d - Trying to remove transactions to make room for certificate (tot_tx: %zu, max: %zu)\n", __func__, __LINE__, totalTxSize, m_max_size / 2);
+        trimToSize(nullptr, current_usage - new_entry_usage);
```
Would it be possible to "optimize" the mempool usage by calling `trimToSize(nullptr, max_size - new_entry_usage);`?

drgora replied:
Makes sense, done

titusen commented on src/wallet/wallet.cpp:
```diff
@@ -3663,7 +3663,7 @@ bool CWallet::AddAccountingEntry(const CAccountingEntry& acentry, CWalletDB & pw
 }
 
-CAmount CWallet::GetMinimumFee(unsigned int nTxBytes, unsigned int nConfirmTarget, const CTxMemPool& pool)
+CAmount CWallet::GetMinimumFee(unsigned int nTxBytes, unsigned int nConfirmTarget, const std::shared_ptr<CTxMemPool> pool)
```
Unnecessary copy constructor call. Please, pass by reference.

drgora replied:
This modifications has been reverted

titusen commented on src/txmempool.cpp:
```diff
@@ -605,9 +658,11 @@ void CTxMemPool::remove(const CTransactionBase& origTx, std::list<CTransaction>&
     for(const uint256& hash : objToRemove)
     {
-        if (mapTx.count(hash))
+        std::map<uint256, CTxMemPoolEntry>::const_iterator entry_it;
```
Please, use shorter syntax:
`       const auto& entry_it = mapTx.find(hash);
        if (entry_it != mapTx.end())`

drgora replied:
Why not, done

JackPiri commented on qa/rpc-tests/mempool_size_limit.py:
```diff
@@ -0,0 +1,593 @@
+        except JSONRPCException as e:
+            assert(False)
+
+        prev_mpool = mpool
```
What about checking the evicted tx is actually the one expected?
After keeping track of tx with minfeerate in previous filling loop, with something like
```
if (feerate < minfeerate):
  minfeerate = feerate
  minfeeratetxid = txid
```
you could then check
```suggestion
prev_mpool = mpool
assert(minfeeratetxid in mpool)
mpool = self.nodes[0].getrawmempool()
assert(minfeeratetxid not in mpool)
```

JackPiri replied:
Or possibly just extracting minimum before the assertions:
`min(feerates, key=feerates.get)`

JackPiri replied:
Agreed to handle this in a separate PR.

JackPiri commented on qa/rpc-tests/mempool_size_limit.py:
```diff
@@ -0,0 +1,593 @@
+            txFile.write(bytes(str(t['amount'] / tsize) + '\n', encoding='utf8'))
+            txFile.write(bytes(signresult["hex"], encoding='utf8'))
+            txFile.close()
+            txs.append({'fee': t['amount'] / tsize, 'hex': signresult["hex"]})
```
I found a way to exactly compute fee rate during data setup
```suggestion
txs.append({'fee': (t['amount'] - 128 * Decimal(0.00005816)) / tsize, 'hex': signresult["hex"]}) # 128 is outputs quantity, 5816 (xb816 reversed) is output value
```
Probably it would be good also to refactor the key name, moving from `'fee'` to `'feerate'`, also to avoid confusion with the key `'fee'` of `create_sc_certificates`, which actually refers to fee itself (not fee rate)
With above computation, I think you can also make simpler the `run_test` function, avoiding to distinguish between `tx_input` and `feerate` (and their corresponding minimum), keeping only fee rate.

JackPiri replied:
Agreed to handle this in a separate PR.

JackPiri commented on src/init.cpp:
```diff
@@ -1090,8 +1091,30 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
+
+    // -mempoollimit limits
+    int64_t nMempoolSizeLimit = GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE_MB) * 1000000;
+    if (nMempoolSizeLimit < 0) {
```
I guess the check here should be consistent with the error message, so:
```suggestion
    if (nMempoolSizeLimit < 4 * 1000000) {
```

drgora replied:
Good catch! Done

JackPiri commented on src/main.h:
```diff
@@ -135,6 +135,8 @@ static const bool DEFAULT_ADDRESSINDEX = false;
 static const bool DEFAULT_TIMESTAMPINDEX = false;
 static const bool DEFAULT_SPENTINDEX = false;
 
+static constexpr unsigned int DEFAULT_MAX_MEMPOOL_SIZE_MB = 400;
```
What about moving this constant to file `txmempool.h`? It seems to me to be more scoped in that way.
I've checked, there should be no building conflict doing this (in case you agree, please double check if you can get rid of now-became-unused includes introduced by this PR, like for example for file `src/test/policyestimator_tests.cpp`).

drgora replied:
I agree with you it would be a great idea. But, unfortunately, right now we have all similar variables (such as `DEFAULT_MAX_ORPHAN_TRANSACTIONS`) defined in `main.h`.
I believe we really should refactor things at some point, and move these variables where they actually belong, in an organized way. But for the time being I'd rather be consistent, and leave it here.

JackPiri commented on src/txmempool.cpp:
```diff
@@ -2009,6 +2071,72 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
+    if (certificate) {
+        LogPrint("mempool", "%s():%d - Trying to remove something to make room for certificate\n", __func__, __LINE__);
+        for (const auto& cert: mapCertificate) {
+            CRawFeeRate feerate = avgFeeRateWithDeps(cert.second.GetCertificate(), true, rates_cache);
```
As a follow-up activity, it can be considered to cache the result even further (in addition to what is already available through `rates_cache`); I'm referring to storing the result in a member field and update it as soon as a tx/cert is added/removed.
Clearly this is valuable if a significant performance improvement can be achieved, otherwise just ignore this comment.

drgora replied:
I see your point, and that was in fact my very first idea. The thing is that any modification in the mempool can potentially have an impact on the entire feerates graph.
In other words, maintaining the graph means doing a complete exploration on each modification. I'm not clear if there is a way to reduce this complexity significantly enough for this idea to bring a performance improvement.
But anyways, we should keep this idea in the radar for the future.

JackPiri replied:
Agreed to handle this in a separate PR.

JackPiri commented on src/gtest/test_mempool.cpp:
```diff
@@ -203,7 +203,7 @@ TEST_F(MempoolTest, TxInputLimit) {
     // UpdateNetworkUpgradeParameters(Consensus::UPGRADE_OVERWINTER, Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT);
 
-    CTxMemPool pool(::minRelayTxFee);
+    std::unique_ptr<CTxMemPool> pool(::minRelayTxFee, DEFAULT_MAX_MEMPOOL_SIZE_MB * 1000000);
```
Ridicolously minor/pedantic comment (line 216 below):
```
 EXPECT_FALSE(AcceptTxToMemoryPool(*pool, state1, tx1, LimitFreeFlag::OFF, &missingInputs, RejectAbsurdFeeFlag::OFF, DisconnectingFlag::OFF,
                                      MempoolProofVerificationFlag::SYNC));
```
(btw, it would not compile anyway :stuck_out_tongue_winking_eye:).

drgora replied:
Done

JackPiri commented on src/rpc/blockchain.cpp:
```diff
@@ -1411,12 +1411,12 @@ UniValue getchaintips(const UniValue& params, bool fHelp)
 UniValue mempoolInfoToJSON()
```
As a follow-up activity, what about providing additional info related to bytes-for-txs and bytes-for-certs?
I know currently in `mempool_size_limit.py` we keep cumulatively track of bytes/sizes, but it could be useful to have those info being required on demand.

drgora replied:
I agree, might be a valuable addition in furure.

JackPiri replied:
Agreed to handle this in a separate PR.

JackPiri commented on qa/rpc-tests/mempool_size_limit.py:
```diff
@@ -0,0 +1,605 @@
+        cert_fees.append(certs[2][0]['fee'] / (len(certs[2][0]['hex']) // 2))
+        cert_fees.append(certs[3][0]['fee'] / (len(certs[3][0]['hex']) // 2))
+
+        mark_logs("Filling mpool with low fee certificates, expecting success and transaction eviction", self.nodes, DEBUG_MODE)
```
Similarly to previous comment, it could be worth checking that at least the lowest fee transaction is evicted (I know possibly the eviction affects multiple transactions, but the lowest fee for sure).

drgora replied:
Good point, but I believe that would require a new data set, or using the current data set in a different way. Currently we might have multiple transactions with the same minimum fee rate, and their ordering is not defined

JackPiri replied:
Agreed to handle this in a separate PR.

JackPiri commented on qa/rpc-tests/mempool_size_limit.py:
```diff
@@ -0,0 +1,605 @@
+            assert_equal(e.error['code'], -7)
+            print("Ok")
+
+        mark_logs("Sending high fee certificates, expecting success and sc0 low quality certificate eviction", self.nodes, DEBUG_MODE)
```
Similarly to previous comment, it could be worth checking that at least the lowest fee certificate is evicted (I know possibly the eviction affects multiple certificate, but the lowest fee for sure).

JackPiri replied:
Agreed to handle this in a separate PR.

JackPiri reviewed (approved):
Wonderful job, this PR is really valuable and clean.
I left some non-blocking comments, it would be nice to have the non-big-effort ones to get addressed in this PR, the others can be considered in a follow-up PR.
Thanks,

titusen commented on src/txmempool.cpp:
```diff
@@ -2009,6 +2071,72 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
+    }
+
+    // we must either reject incoming tx, or remove something
+    int64_t size_to_be_removed = current_usage + new_entry_usage - max_size;
```
Please, change these to std::unordered_map. It has better operational complexity.

drgora replied:
I agree, done with the latest set of updates.

titusen reviewed (approved):
Besides follow-up LGTM. Thanks.

drgora committed:
Simplify remove() calls in txmempool

drgora committed:
Use unordered data structures when indexing on uint256

drgora committed:
Add dryrun option to trimToSize

This is used to perform all the mempool size limit checks on certificates before verifying certificates proofs.

drgora committed:
Simplify mempool_size_limit Python test

drgora committed:
Add more detailed info to mempoolInfoToJSON

Specifically, this commit adds info about how many bytes in mempool are consumed by transactions and certificates.

ptagl reviewed (approved):
Looks good to me, thanks!
